### ISSUE TITLE
0.63.14 - big troubles with SevenZipSharp

### ISSUE THREAD

**Issue by siaccarino** [Nexus-Mods/Nexus-Mod-Manager#95]
It take ages to unpack huge Skyrim (oldrim) mods (>500mb) - even on decent hardware
It constantly fails installing huge mods (>500mb)
Error message is always about a problem with SevenZipSharp

**Comment by elgonzo** [comment]
Yeah, other problems with the SevenZipSharp library not withstanding, the slow uncompressing of 7z archives could be caused by NMM unpacking the files from the archive in the wrong order.

7z archives usually aggregates several files in so-called solid blocks which will then be compressed as one LZMA stream. Unpacking of these files should be done in the order in which they appear in the block. Unpacking in a different order comes with a performance penalty, as the uncompressor cannot freely jump around in the block but can only sequentially decompress it.

When the decompressor is ordered to decompress a file that is located before the current position in the LZMA stream, the compressor has no choice but starting to decompress the LZMA stream from the beginning, again. Which, if larger 7z archives are involved, can quickly become very costly...

(I don't know why the NMM team decided to use SevenZipSharp. On the other hand, 7z support in .NET is spotty; it's pretty much impossible to find a full-fledged, robust, up-to-date 7z library out there. I was working once on a .NET project of mine that required 7z read support. I eventually resorted to interface with the 7z.dll directly and not use a 3rd-party library, since i wasn't able to find one that didn't crap out this or that way. Needless to say that my .NET wrapper for 7z.dll is far from being full-fledged or robust, and probably being pretty crappy as well. Go figure :) )

**Comment by ousnius** [comment]
The biggest problem with this is that NMM extracts all *.txt files BEFORE the actual installation of the mod into a separate directory. So if a 500 MB archive has 8 text files, it's gonna stream the whole thing 8 times before even installing it. This is an option in the settings, turned on by default.

**Comment by elgonzo** [comment]
> The biggest problem with this is that NMM extracts all *.txt files BEFORE the actual installation of the mod into a separate directory. So if a 500 MB archive has 8 text files, it's gonna stream the whole thing 8 times before even installing it. This is an option in the settings, turned on by default.

Ugh... unless it is guaranteed that the txt files are either at the beginning of a solid block or are packed in their own little block  (how could the NMM devs guarantee this?) , this is just terrible with respect to larger mods...

**Comment by squid-box** [comment]
Tagging this as an enhancement, but I doubt anyone will look into this now that NMM is no longer supported by Nexus.
